lib/Graph/TensorLayout.cpp

```diff
@@ -0,0 +1,277 @@
+    // layout.
+    if (auto *layout = getLayoutFromName(TN->getLayout())) {
+      return *layout;
+    }
+    // Dynamically form the layout description for transposes.
+    auto input = TN->getInput();
+    auto inputLayout = getNthInputLayoutRequirements(node, 0);
+    llvm::SmallVector<TensorDimensionDescription, max_tensor_dimensions> dims(
+        input.dims().size());
+    auto shuffle = TN->getShuffle();
+    for (unsigned idx = 0, e = inputLayout.getNumDims(); idx < e; ++idx) {
+      dims[shuffle[idx]] = inputLayout.getNthDimDescription(idx);
+    }
+    TensorLayoutDescription tld(dims);
+    return tld;
+  }
+  if (auto *C = llvm::dyn_cast<Constant>(node)) {
+    if (auto *layout = getLayoutFromName(C->getLayout())) {
+      return *layout;
+    }
+  }
+  if (auto *PH = llvm::dyn_cast<Placeholder>(node)) {
+    if (auto *layout = getLayoutFromName(PH->getLayout())) {
+      return *layout;
+    }
+  }
+  if (auto *RN = llvm::dyn_cast<ReshapeNode>(node)) {
+    if (auto *layout = getLayoutFromName(RN->getLayout())) {
+      return *layout;
+    }
+    auto result = node->getNthResult(n);
+    auto *user = (*result.getUsers().begin()).getUser();
+    int inputIdx = getInputIdx(user, result);
+    if (inputIdx >= user->getNumInputs() || llvm::isa<TransposeNode>(user)) {
+      return getLayoutsForDims()[dims.size()];
+    }
+    auto layout = getNthInputLayoutRequirements(user, inputIdx);
+    if (auto *layoutDesc = getLayoutFromName(layout.getLayoutName())) {
+      return *layoutDesc;
+    }
+  }
+  return getLayoutsForDims()[dims.size()];
+}
+
+static const TensorLayoutDescription
+returnBaseReqOrNHWC(TensorLayoutDescription &baseReq) {
+  if (!baseReq.isSameLayout(
+          CanonicalTensorLayout::getInstance().getLayoutsForDims()[4])) {
+    return baseReq;
+  }
+  // NHWC is the canonical default
+  return CanonicalTensorLayout::getInstance().getDefaultNDLayout(4);
+}
+
+const TensorLayoutDescription
+CanonicalTensorLayout::getNthInputLayoutRequirements(const Node *node,
+                                                     size_t n) const {
+  auto baseReq = TensorLayoutCommon::getNthInputLayoutRequirements(node, n);
+  if (acceptsAnyLayout(node)) {
+    return baseReq;
+  }
+  return returnBaseReqOrNHWC(baseReq);
+}
+
+const TensorLayoutDescription
+CanonicalTensorLayout::getNthResultLayoutRequirements(const Node *node,
+                                                      size_t n) const {
+  auto baseReq = TensorLayoutCommon::getNthResultLayoutRequirements(node, n);
+  return returnBaseReqOrNHWC(baseReq);
+}
+
+const TensorLayoutDescription
+CanonicalTensorLayout::getDefaultNDLayout(unsigned dims) const {
+  if (dims == 4) {
+    return layoutNHWC;
+  }
+  return TensorLayoutCommon::getDefaultNDLayout(dims);
+}
+
+bool CanonicalTensorLayout::acceptsAnyLayout(const Node *node) const {
+  switch (node->getKind()) {
+  case Kinded::Kind::AddNodeKind:
```

Comment:
I think that we have a flag that checks if an operation is an element-wise operation.